Source/Dafny/Compilers/Compiler-python.cs

```diff
@@ -886,43 +890,63 @@ protected class ClassWriter : IClassWriter {
 
       switch (op) {
         case BinaryExpr.ResolvedOpcode.And:
-          opString = "and"; break;
+          opString = "and";
+          break;
 
         case BinaryExpr.ResolvedOpcode.Or:
-          opString = "or"; break;
+          opString = "or";
+          break;
 
         case BinaryExpr.ResolvedOpcode.LeftShift:
-          opString = "<<"; break;
+          opString = "<<";
+          break;
 
         case BinaryExpr.ResolvedOpcode.RightShift:
-          opString = ">>"; break;
+          opString = ">>";
+          break;
 
         case BinaryExpr.ResolvedOpcode.Add:
-        case BinaryExpr.ResolvedOpcode.Concat:
           if (!resultType.IsCharType) {
+            truncateResult = true;
             opString = "+";
           } else {
             staticCallString = "_dafny.PlusChar";
           }
           break;
 
+        case BinaryExpr.ResolvedOpcode.Concat:
+          opString = "+";
+          break;
+
         case BinaryExpr.ResolvedOpcode.Sub:
           if (!resultType.IsCharType) {
+            truncateResult = true;
             opString = "-";
           } else {
             staticCallString = "_dafny.MinusChar";
           }
           break;
 
         case BinaryExpr.ResolvedOpcode.Mul:
-          opString = "*"; break;
+          opString = "*";
+          truncateResult = true;
+          break;
 
-        //Fixme: Ints
         case BinaryExpr.ResolvedOpcode.Div:
-          opString = "/"; break;
+          if (resultType.IsIntegerType || resultType.IsBitVectorType || resultType.AsNewtype != null) {
+            staticCallString = "_dafny.euclidianDivision";
+          } else {
+            opString = "/";
+          }
+          break;
 
         case BinaryExpr.ResolvedOpcode.Mod:
-          opString = "%"; break;
+          if (resultType.IsIntegerType || resultType.IsBitVectorType || resultType.AsNewtype != null) {
+            staticCallString = "_dafny.euclidianModulus";
+          } else {
+            opString = "%";
```

Comment:
What case does this correspond to? Real numbers?

Reply:
Good point, I probably added it for symmetry in `/`.